app/controllers/follows_controller.rb

```diff
@@ -3,19 +3,25 @@ class FollowsController < ApplicationController
 
   def show
     skip_authorization
-    unless current_user
-      render plain: "not-logged-in"
-      return
-    end
+    render(plain: "not-logged-in") && return unless current_user
+
     if current_user.id == params[:id].to_i && params[:followable_type] == "User"
       render plain: "self"
       return
-    elsif params[:followable_type] == "User" && FollowChecker.new(current_user, params[:followable_type], params[:id]).cached_follow_check && FollowChecker.new(User.find(params[:id]), params[:followable_type], current_user.id).cached_follow_check
+    end
+
+    following_them_check = FollowChecker.new(current_user, params[:followable_type], params[:id]).cached_follow_check
+
+    return render plain: following_them_check unless params[:followable_type] == "User"
+
+    following_you_check = FollowChecker.new(User.find_by(id: params[:id]), params[:followable_type], current_user.id).cached_follow_check
+
+    if following_them_check && following_you_check
       render plain: "mutual"
-    elsif params[:followable_type] == "User" && FollowChecker.new(User.find(params[:id]), params[:followable_type], current_user.id).cached_follow_check
+    elsif following_you_check
```

Comment:
I think there's still one case where there's one unnecessary computation, and it's this.
If you look at the left column this case is "type is user and following you check", which means that "following_them" has been computed but it's never used in this case

Reply:
No matter what I believe once you have established it is a "User" you need the following them check in order to determine if it is mutual or not. Here it has been used to verify that the following is not mutual. Even if we swtiched this boolean we would have to validate it was false before showing "follow-back"

Reply:
I tried to rewrite the current code textually because I didn't fully get it:

```text
1. If the current user ID and the given user ID are the same AND the type is User RENDER self
2. else if the type is User AND you are following them AND they are following you RENDER mutual
3. else if the type is User AND they are following you RENDER follow-back
4. else check if you're following them
```

This code is doing multiple things but what we can understand here is:

- if the type is not User we need to check if you're following them (this might be a logic issue, but still, it's how the code works)
- if the type is user we need to check the following:
  - MUTUAL: if you follow_them AND if they follow you (but the second one is not needed, there's no reason to check the "mutualness" if the first check is false)
  - FOLLOW BACK: here we're only interested if they are following the current user

I re-ran all this logic in my head and I agree :-) 

Maybe we should add a paragraph of explanation for future readers :D